### PR TITLE
firmware/raspberrypi: Add a get_throttled sysfs file

### PR DESCRIPTION
**Pull request by notro**
This is a result of the discussion in #2367.

Kernel log:
```
[  530.474085] Under-voltage detected! (0x00050005)
[  534.634106] Voltage normalised (0x00000000)
```

A simple script that polls on the sysfs file and shows the changes:
```python
import select
import datetime

epoll = select.epoll()

file = open("/sys/devices/platform/soc/soc:firmware/get_throttled")
epoll.register(file.fileno(), select.EPOLLPRI | select.EPOLLERR)
status = file.read()

while(True):
    epoll.poll()
    file.seek(0)
    status = file.read()
    get_throttled = int(status, 16)
    print("%s 0x%08x" % (datetime.datetime.now().strftime('%H:%M:%S'), get_throttled))

epoll.unregister(file.fileno())
file.close()

```

A simple proof of concept for desktop notification:
```python
import dbus
import select
import datetime
import signal

msg = """Under-voltage is blah blah not good.
See: http://some-explanation.org
"""

def quit(signo, _frame):
    exit(0)

for sig in ('TERM', 'HUP', 'INT'):
    signal.signal(getattr(signal, 'SIG'+sig), quit);

bus = dbus.SessionBus()
notifier = bus.get_object('org.freedesktop.Notifications', '/org/freedesktop/Notifications')
iface = dbus.Interface(notifier, 'org.freedesktop.Notifications')
nid = None

epoll = select.epoll()
file = open("/sys/devices/platform/soc/soc:firmware/get_throttled")
epoll.register(file.fileno(), select.EPOLLPRI | select.EPOLLERR)
status = file.read()

while(True):
    epoll.poll()
    file.seek(0)
    status = file.read()
    get_throttled = int(status, 16)

    if get_throttled & (1 << 16):
        try:
            nid = iface.Notify('get_throttled', 0, 'dialog-warning', 'Under-voltage!', msg, [], {}, 0)
        except dbus.exceptions.DBusException as e:
            sys.stderr.write("Error sending notification: %s\n" % e); sys.stderr.flush()
    elif nid:
        try: # The user might have closed it
            iface.CloseNotification(nid)
        except dbus.exceptions.DBusException:
            pass
        nid = None

epoll.unregister(file.fileno())
file.close()

```

dbus notification daemon (supports cancelling notifications):
```
sudo apt install dunst
```

### CODE REVIEW

File: drivers/firmware/raspberrypi.c [hunk 1]
```diff
@@ -166,6 +170,101 @@ int rpi_firmware_property(struct rpi_firmware *fw,
...
+	old_uv = old_sticky & UNDERVOLTAGE_BIT;
+
+	if (new_uv != old_uv) {
+		if (new_uv)
+			pr_crit("Under-voltage detected! (0x%08x)\n", *value);
+		else
+			pr_info("Voltage normalised (0x%08x)\n", *value);
+	}
+
+	sysfs_notify(&fw->cl.dev->kobj, NULL, "get_throttled");
```
Comment (pelwell) [inline drivers/firmware/raspberrypi.c:224]: I know the polling rate is very low so the answer is largely moot, but is there a reason for waking any waiters regardless of whether the state has changed?

Comment (notro) [reply drivers/firmware/raspberrypi.c:224]: We only reach this if there are changes, see a couple of lines up:
```c
if (new_sticky == old_sticky)
	return 0;
```

**Comment by pelwell** [comment]
No objections.

**Comment by cyr-ius** [comment]
I compiled the last kernel 4.14.27 and since then I have the following message in loop in the kern.log
raspberrypi-firmware soc:firmware: Request 0x00030046 returned status 0x80000001
and i have no problem of under voltage
Pi Zero W
vcgencmd measure_clock arm
frequency(45)=1000106000
vcgencmd measure_clock core
frequency(1)=400000000
vcgencmd measure_volts
volt=1.2000V

vcgencmd measure_temp; vcgencmd get_throttled; vcgencmd measure_temp
temp=37.9'C
throttled=0x0
temp=37.9'C

I have the problem on my Pi Zero and Pi 2

How to stop this message ?

**Comment by pelwell** [comment]
Answered elsewhere.

**Comment by jcea** [comment]
I am using kernel "4.19.122-1-osmc" in a Raspberry Pi 3B+ Rev 1.3. The file "/sys/devices/platform/soc/soc:firmware/get_throttled" exists, but it always returns zero:

```
root@osmcpi:~# vcgencmd get_throttled
throttled=0x80008
root@osmcpi:~# cat "/sys/devices/platform/soc/soc:firmware/get_throttled" 
0
```

More details in https://discourse.osmc.tv/t/sys-devices-platform-soc-soc-firmware-get-throttled-not-working/86639

**Pull request merged**